### PR TITLE
Implement cancel button for the sidebar

### PR DESCRIPTION
**Pull request by tsmacdonald**
This cancel button (and its friends for emails and new pulses) closes the sidebar now:

image

### CODE REVIEW

File: frontend/src/metabase/dashboard/components/Dashboard.jsx [hunk 1]
```diff
@@ -228,6 +230,10 @@ export default class Dashboard extends Component {
     });
   };
 
+  onCancel = () => {
+    this.setState({ isSharing: false });
```
Comment (paulrosenzweig) [inline frontend/src/metabase/dashboard/components/Dashboard.jsx:234]: It's not part of this PR, but `isSharing` might make more sense in Redux rather than component state. I think it hinges on whether we want to tie the sidebar behavior in with any other UI changes that are managed in redux. e.g. if we the sharing sidebar to auto-close when switching to editing mode, we can more cleanly handle that with reducers.

File: frontend/src/metabase/sharing/components/SharingSidebar.jsx [hunk 2]
```diff
@@ -561,7 +561,7 @@ class SharingSidebar extends React.Component {
       const slackSpec = formInput.channels.slack;
 
       return (
-        <Sidebar onCancel={() => true}>
+        <Sidebar onCancel={onCancel}>
```
Comment (paulrosenzweig) [inline frontend/src/metabase/sharing/components/SharingSidebar.jsx:564]: 👍 threading it through like this makes sense.

Separately, we might be able to consolidate the separate but identical `<Sidebar>` uses here, but that's outside of this PR.

**Review by paulrosenzweig — COMMENTED** [review]
👏 woohoo first frontend PR on Metabase!

File: frontend/src/metabase/dashboard/components/Dashboard.jsx (hunk 1 above)
Comment (tsmacdonald) [reply frontend/src/metabase/dashboard/components/Dashboard.jsx:234]: @paulrosenzweig This sounds really appealing to me—there's already some funky behavior surrounding getting things closed. I spent a while exploring the actions, `Dashboard` state, etc. and didn't make too much headway. Do you have 10-15 minutes in the next few days to talk through this a little more?

Comment (robdaemon) [reply frontend/src/metabase/dashboard/components/Dashboard.jsx:233]: Ahhh cool I'd like a better explanation of this too - can I join the talk about it? I was looking at how other things were done like this using `props` instead of `state` but I couldn't figure out how that worked since `props` are supposed to be immutable?

There's clearly a redux-ism here I'm missing!

File: src/metabase/models/pulse.clj [hunk 3]
```diff
@@ -240,24 +240,18 @@
...
+                     :from      [[Pulse :p]]
+                     :where     [:and
+                                 [:= :p.alert_condition nil]
+                                 [:= :p.archived archived?]]
+                     ;; :order-by  [[:%lower.name :asc]]
+                     }
+         query      (merge-with conj base-query (when dashboard_id
+                                                  {:join      [[PulseCard :pc]     [:= :p.id :pc.pulse_id]]
+                                                   :left-join [[DashboardCard :dc] [:= :pc.dashboard_card_id :dc.id]]
+                                                   :where     [:= :dc.dashboard_id dashboard_id]}))]
```
Comment (robdaemon) [inline src/metabase/models/pulse.clj:254]: Does this actually merge properly, or do you end up with:

`[:and [:= :p.alert_condition nil]
         [:= :p.archived archived?]
 := :dc.dashboard_id dashboard_id]`

?

Comment (tsmacdonald) [inline src/metabase/models/pulse.clj:254]: @robdaemon Looks like you answered your own question re. `merge-with conj`. I confess that I don't understand SQL well enough to know why the `distinct` is necessary; I would've thought the query as-is wouldn't return duplicate results.

File: frontend/src/metabase/dashboard/components/Dashboard.jsx [hunk 4]
```diff
@@ -228,8 +229,12 @@ export default class Dashboard extends Component {
     });
   };
 
+  onCancel = () => {
+    this.props.setSharing(false);
+  };
+
   onSharingClick = () => {
-    this.setState({ isSharing: !this.state.isSharing });
+    this.props.setSharing(true);
```
Comment (tsmacdonald) [inline frontend/src/metabase/dashboard/components/Dashboard.jsx:237]: I couldn't see any cases where the toggling was actually necessary

File: frontend/src/metabase/dashboard/components/Dashboard.jsx [hunk 5]
```diff
@@ -331,7 +337,11 @@ export default class Dashboard extends Component {
                   )}
                 </div>
               </div>
-              <Sidebars {...this.props} isSharing={isSharing} />
+              <Sidebars
+                {...this.props}
+                isSharing={isSharing}
+                onCancel={this.onCancel}
+              />
```
Comment (tsmacdonald) [inline frontend/src/metabase/dashboard/components/Dashboard.jsx:344]: Tbh it's still a little perverse to me that I need to thread both the state (`isSharing`) and its mutator (`onCancel`) through, but I can roll with it. I think it'd be more compelling if the data was more complex than a single boolean value.

File: frontend/src/metabase/dashboard/components/DashboardActions.jsx [hunk 6]
```diff
@@ -8,7 +8,7 @@ import PopoverWithTrigger from "metabase/components/PopoverWithTrigger";
 import RefreshWidget from "metabase/dashboard/components/RefreshWidget";
 import Tooltip from "metabase/components/Tooltip";
 
-export const getDashboardActions = ({
+export const getDashboardActions = (self, {
```
Comment (tsmacdonald) [inline frontend/src/metabase/dashboard/components/DashboardActions.jsx:11]: I feel gross about this. I tried to use `getDashboardActions.bind(this)(this.props)` where this is used but it didn't work and I maybe don't grok `bind`

Comment (paulrosenzweig) [reply frontend/src/metabase/dashboard/components/DashboardActions.jsx:11]: I agree on the gross feels. Could you extract the `PopoverWithTrigger` code below to its own class-based component?

**Review by paulrosenzweig — COMMENTED** [review]
Looking good!

File: frontend/src/metabase/dashboard/components/Dashboard.jsx (hunk 5 above)
Comment (tsmacdonald) [reply frontend/src/metabase/dashboard/components/Dashboard.jsx:344]: oh, d'oh, good call

**Comment by tsmacdonald** [comment]
Hey @paulrosenzweig, I addressed all your comments except refactoring `getDashboardActions` / the refs. I added that to a tech debt section on [the hub issue](https://github.com/metabase/metabase/issues/13671) and think it'd be more expedient to circle back to it once I've done a bit more of the feature work, because I think I'll do a better job with a little more experience, I'm worried about the feature work, and it's entirely possible the surrounding code will need to get changed as part of the feature work anyway.

So is it cool to merge this (into `subscribe-to-dashboards-oss`, not `master`)?

The build fails upstream too. Dealing with that is next on the list

**Review by paulrosenzweig — APPROVED** [review]
👍 merging now sounds good to me!

**Review by model2020 — COMMENTED** [review]
Yes
